src/minefield.c

```diff
@@ -93,14 +93,16 @@ bool minefield_reveal_tile(struct Minefield *minefield, size_t x, size_t y) {
     // TODO: this is kinda ugly
     size_t x_end = x < minefield->width - 1 ? x + 1 : x;
     size_t y_end = y < minefield->height - 1 ? y + 1 : y;
+    bool no_mines = true;
     for (size_t x1 = x_start; x1 <= x_end; x1++) {
         for (size_t y1 = y_start; y1 <= y_end; y1++) {
-            if (!minefield_get_tile(minefield, x1, y1)->visible) {
-                minefield_reveal_tile(minefield, x1, y1);
+            struct Tile *surtile = minefield_get_tile(minefield, x1, y1);
+            if (!surtile->visible && !surtile->flagged) {
+                no_mines &= minefield_reveal_tile(minefield, x1, y1);
```

Comment:
This feels too magical to me; I personally would've just done an early `return false;` here and then a `return true` outside the loops. But I don't feel like dealing with changing it so I'll keep it.